### PR TITLE
Papercut typing fixes #3022 jmatthews

### LINKED ISSUE #3022: Improve typing across codebase
## Description of issue / requirement to address

We should improve typing across the codebase with the eventual aim of turning mypy on in our code qa tests.

## Proposed solution

**Slowly** make PRs solving ~100 type errors from running 
`mypy --config-file pyproject.toml bluemira`

## Alternative solutions

Have bad typing forever...

### Stylistic points:

- No implicit None return types:

bad
```python
def a(var: str) -> None:
     print(var)
```
good
```python
def a(var: str):
    print(var)
```
- Use `numpy.typing` for arrays:

```python
import numpy.typing as npt

def a(var: npt.NDArray[float]):
    ...
```
Unfortunately this is not valid according to mypy therefore we have to use the overly restrictive `np.float64` or possibly build a `np.floating` type but for now use the below:

```python
import numpy.typing as npt

def a(var: npt.NDArray[np.float64]):
    ...
```
there is also `npt.ArrayLike` if a function takes for instance either floats or arrays

- Use of `Any` and `object` typing should be avoided these are very generic.
- Types of variables to functions can be loose but types returned from functions should be explicit and confined
### List of PRs
- [x] #3053 
- [x] #3092 
- [x] #3186 
- [x] #3267
- [x] #3305 
-  [ ] #3975

### PR DESCRIPTION
**Pull request by kj5248**
## Linked Issues

Issue #3022 on fixing typing in codebase

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/magnetostatics/biot_savart.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 from typing import TYPE_CHECKING
 
 import numpy as np
+import numpy.typing as npt
```
Comment (je-cook) [inline bluemira/magnetostatics/biot_savart.py:16]: Its already in the TYPE_CHECKING block
```suggestion
```

File: bluemira/magnetostatics/circuits.py [hunk 2]
```diff
@@ -33,10 +35,16 @@ class PlanarCircuit(SourceGroup):
...
     current: float
 
-    def _generate_sources(self, shape, current, source_class, xs_args):
+    def _generate_sources(
+        self,
+        shape: npt.NDArray[np.float64] | Coordinates,
+        current: float,
+        source_class: Any,
+        xs_args: npt.NDArray[np.float64],
+    ) -> list[Any]:
```
Comment (je-cook) [inline bluemira/magnetostatics/circuits.py:47]: Any is bad don't use Any where possible, is there either a abstract baseclass or a protocol we can create?

File: bluemira/utilities/opt_variables.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from typing import TYPE_CHECKING, TextIO, TypedDict
 
 import numpy as np
+import numpy.typing as npt
```
Comment (je-cook) [inline bluemira/utilities/opt_variables.py:19]: Move to TYPE_CHECKING block

**Review by je-cook — CHANGES_REQUESTED** [review]
One question about axis and Iterable and one about the use of Any, otherwise looks good

File: bluemira/geometry/parameterisations.py [hunk 4]
```diff
@@ -1350,7 +1372,7 @@ def _make_domed_leg(
         z_top: float,
         z_mid: float,
         ri: float,
-        axis: Iterable[float] = (0, -1, 0),
+        axis: tuple[float, float, float] = (0, -1, 0),
```
[the remark this answers is not in the document]
Comment (kj5248) [reply bluemira/geometry/parameterisations.py:1375]: When you use Iterable there is a typing mismatch to other files and functions. Wasn't sure whether to change those to Iterable or this to tuple

[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/geometry/parameterisations.py:1375]: I think they should be Iterable too as I think the underlying matplotlib function is agnostic

File: bluemira/magnetostatics/circuits.py (hunk 2 above)
Comment (kj5248) [reply bluemira/magnetostatics/circuits.py:47]: Yeah I tried a couple of things but got a variety of errors. Do I have to go all the way to the baseclass for that?

Comment (je-cook) [reply bluemira/magnetostatics/circuits.py:47]: So it need to include every possible source class (eg the base class or a common ancestor) or a protocol which implements stubs of the methods a source will have (see your next papercut for an example)

Comment (kj5248) [reply bluemira/magnetostatics/circuits.py:47]: So when I use a common ancestor for PolyhedralPrismCurrentSource and TrapezoidalPrismCurrentSource which is CrossSectionCurrentSource it throws the error that it is not callable and also that "Type[TrapezoidalPrismCurrentSource]" doesn't match "CrossSectionCurrentSource"

Comment (je-cook) [reply bluemira/magnetostatics/circuits.py:47]: You need as the type `type[CrossSectionCurrentSource]` because you are passing in an unitialised class

Comment (kj5248) [reply bluemira/magnetostatics/circuits.py:47]: Ah I was using type() before. That makes a difference, although it seems to have a weird issue...  error: "TrapezoidalPrismCurrentSource" gets multiple values for keyword argument "alpha",

Comment (je-cook) [reply bluemira/magnetostatics/circuits.py:47]: resolved offline

**Review by je-cook — APPROVED** [review]

**Pull request merged**